helix-core/src/uri.rs

```diff
@@ -0,0 +1,102 @@
+use std::path::{Path, PathBuf};
```

Comment:
I think this fits as a core data type similar to `helix-core/src/diagnostic.rs` - it's what Helix understands as a URI and not necessarily a useful URI type for other applications. I think it depends on how we want to use stdx.

There's a difference between https://matklad.github.io/2021/05/12/design-pattern-dumping-ground.html and rust-analyzer's stdx crate - my interpretation of the blog post is that there should be a kitchen-sink place to put all code that doesn't have a natural crate, but rust-analyzer's stdx crate focuses on being extensions to stdx that don't overlap with rust-analyzer and have minimal dependencies (i.e. [this line](https://github.com/rust-lang/rust-analyzer/blob/14924b03ce5dd9b64cebc5c06a7484ea6516f9ac/crates/stdx/Cargo.toml#L21)).

I think it makes sense for Helix-specific data types that are generic like diagnostics and URIs to live in core while generic 'things that help us write Rust' like https://github.com/helix-editor/helix/blob/87266b0a07af4b1a078d07fc0cbd11a6e08a1083/helix-config/src/definition/lsp.rs#L38-L56 fit really well into stdx.

What do you think? I like the "use stdx only for extensions to the stdlib" interpretation but I don't feel strongly about it - the kitchen sink also seems pretty useful to me

Reply:
+1 for `helix-core` since a URI type, modelled exclusively on which schemes *helix supports* feels unsuited for an `*-stdx` crate.

it's even possible that URI might end up being a unrepresentative name for this type, and what we want is more akin to "document identifier", which is even more integral.

note that in a project like helix, the kitchen sink solution would be a bit more fragmented since we'd need to support both things at the top of dependency graph (where `helix-stdx` is) and closer to the bottom, i.e. we can't just have a single `helix-kitchen-sink` crate.

Reply:
I do want stdx to be a kind of kitchen sink like in that blog post. I think rust-analyzer actually is like that to just most of it is std extensions. If stdx doesn't become that kind of kitchen sink that core will (or its littered troughout the codebas) which is rtsictly worse. For example the rope extension traits is not really strictly speaking an std extension but I don't think it belongs  inc core since its not in any way helix specific just some utility functions that could be build into upstream ropey but aren't.

I do think you mak a good point about the URI type in this case. The exact variants will reflect functionality we support so its helix specific and belongs in core (that way it could even integrate with plugins in the future). If this was a generic URI/URL implementation like in the `url` crate then I would place it in helix-stdx.